### PR TITLE
Review: feat(comment): add an AST node for the javadoc tags

### PR DESCRIPTION
**Pull request by tdurieux**
This one is for you @msteinbeck

### CODE REVIEW

**Comment by msteinbeck** [comment]
@tdurieux, you are awesome :).

**Comment by tdurieux** [comment]
http://spoon.gforge.inria.fr is dead. The build will ne pass.

**Comment by surli** [comment]
It seems ok for me, I'm waiting for http://spoon.gforge.inria.fr to be up again, and when checks will be ok I'll merged.

**Comment by msteinbeck** [comment]
According to oracle, the first sentence of a JavaDoc string is referred as short summary. For instance:

```java
/**
  * This is a short summary. Now following the long description.
  */
...
```

It might be useful to add getters for the short summary and long description to the AST node.

**Comment by tdurieux** [comment]
Does the  long description contain the  short summary?

**Comment by msteinbeck** [comment]
> Does the long description contain the short summary?

Oracle says: "Write the first sentence as a short summary of the method, as Javadoc automatically places it in the method summary table (and index)." (http://www.oracle.com/technetwork/articles/java/index-137868.html)

So it looks like it's up to us defining how to handle the summary. I would suggest the following:
1. A JavaDoc string provides a description. It contains anything except of tags.
2. If desired, one can fetch the summary using a getter.
3. If desired, one can fetch the long description without the summary using a getter.

**Comment by tdurieux** [comment]
> "Write the first sentence
Is it the first sentence or the first line?

**Comment by msteinbeck** [comment]
> Is it the first sentence or the first line?

As far as I know it's the first sentence delimited using a "." (dot). Consequently, sentences like: "The is a short summary, i.e., ..." lead to a _wrong_ summary, as JavaDoc creates the following summary: "The is a short summary, i."

**Comment by tdurieux** [comment]
For me is done

File: src/main/java/spoon/reflect/code/CtJavaDoc.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	 * Remove a specific tag
+	 * @param tag the tag to remove
+	 */
+	<E extends CtJavaDoc> E removeTag(CtJavaDocTag tag);
+
+	/**
+	 * Get the sort summary of the javadoc (first sentence of the javadoc)
+	 * @return the summary of the javadoc
+	 */
+	String getSortDescription();
```
Comment (surli) [inline src/main/java/spoon/reflect/code/CtJavaDoc.java]: typo: `getShortDescription`

Comment (surli) [inline src/main/java/spoon/reflect/code/CtJavaDoc.java]: (also in the comment)

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 2]
```diff
@@ -799,6 +801,36 @@ public void visitCtDo(CtDo doLoop) {
...
+		}
+
+		String[] tagLines = docTag.getContent().split(EOL);
+		for (int i = 0; i < tagLines.length; i++) {
+			String com = tagLines[i];
+			if (i > 0 || docTag.getType().hasParam()) {
+				printer.write(" * ");
+			}
+			if (docTag.getType().hasParam()) {
+				printer.write("\t\t");
```
Comment (surli) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: Don't we have any constant for that?

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 3]
```diff
@@ -799,6 +801,36 @@ public void visitCtDo(CtDo doLoop) {
...
 	@Override
+	public void visitCtJavaDoc(CtJavaDoc comment) {
+		visitCtComment(comment);
+	}
+
+	@Override
+	public void visitCtJavaDocTag(CtJavaDocTag docTag) {
+		String EOL = System.getProperty("line.separator");
+
+		printer.write(" * ");
```
Comment (surli) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: Same here: constant?

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 4]
```diff
@@ -840,6 +873,14 @@ public void visitCtComment(CtComment comment) {
 				}
 
 			}
+			if (comment instanceof CtJavaDoc) {
+				if (!((CtJavaDoc) comment).getTags().isEmpty()) {
+					printer.write(" *").writeln().writeTabs();
```
Comment (surli) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: There's a reason to not put a space after the * here?

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java (hunk 2 above)
Comment (tdurieux) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: Yes here: JavaDocTest#testJavaDocReprint

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java (hunk 4 above)
Comment (tdurieux) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: Yes the current contract: JavaDocTest#testJavaDocReprint

**Pull request merged**